lib/generators/rspec/scaffold/templates/index_spec.rb

```diff
@@ -18,8 +18,9 @@
 
   it "renders a list of <%= ns_table_name %>" do
     render
+    cell_selector = Rails::VERSION::STRING >= '7' ? 'div>p' : 'tr>td'
```

Comment:
I suggest moving this inside `<% %>` so it doesn't appear in the resulting spec.

Reply:
That won't help much.  Something needs to appear in the resulting spec, and that something is going to be different for Rails 7.0 and for previous versions of Rails.  So either we duplicate the logic in the spec (which seems kind of pointless) or we put it in a single variable here.

Reply:
I imagine the flow as:
1. generate a scaffold
2. upgrade Rails from 6 to 7

I can't imagine a failure if we've hardcoded `tr>td`. Because both the `index_spec` and the `index` have it like that being generated by Rails 6.


But, if we keep this ternary, after Rails 7 upgrade, `index_spec` will start failing, because `index` has `td` inside `tr`, while the spec would attempt to select `div>p`.

Reply:
No, that's exactly backwards.

The scaffold is generated dynamically as part of the spec run.  With the version of Rails that is currently active.  So for you get:

Rails 6.x - tr>td
Rails 7.x - div>p

Hardcoding either one in the scaffold spec causes the scaffold spec to fail in at least one case.  

This is easy to confirm on the results on the current branch.  With the current code the scaffold and the various index specs pass across all Ruby/Rails combos.  Hardcoding will break it.

Reply:
> scaffold is generated dynamically as part of the spec run

You think RSpec CI, I mean a real app developer's local machine.

RSpec CI would be green both ways. Real app developer's local test run will fail with the way it is now.

Please don't get me wrong, I suggest to:
```ruby
  it "renders a list of <%= ns_table_name %>" do
    render
    
<% for attribute in output_attributes -%>
    assert_select <% Rails::VERSION::STRING >= '7' ? 'div>p' : 'tr>td' %>, text: Regexp.new(<%= value_for(attribute) %>.to_s), count: 2
<% end -%>
  end
```

On Rails 6, it will generate a spec with lines like:
```ruby
    assert_select 'tr>td', text: Regexp.new(3.to_s), count: 2
```

I think given the HTML table template Rails 6 used to generate, on developer's machine after update to Rails 7 this code will remain working.

What makes this code problematic?

Reply:
> we've hardcoded

Sorry for being ambiguous.
By we I didn't mean me or you, I meant the template, `lib/generators/rspec/scaffold/templates/index_spec.rb`.
And by the product that will undergo hardcoding I meant the resulting `index_spec.rb` spec file on the real project's developer machine, not the template in this git repository.

Reply:
So the suggested approach doesn't work.  It breaks in CI and breaks on a developer's box.

* CI - for Rails 7 your approach will work for the generated index spec you reference (because the `tr>td` will be replaced in the generated spec.  But the scaffold spec - `spec/generators/rspec/scaffold/scaffold_generator_spec.rb ` - will fail, because it will have a hard-coded `tr>td`

* Developer's Box - We have the same issue for the scaffold spec if we generate the app at one Rails version and switch to another.  If a developer builds the smoke app under Rails 7, then the specs now fail.  Why does Rails 6 have a preferential position over Rails 7?  What happens if I generate the app on Rails 7 and move to Rails 6.

When a developer changes RAILS_VERSION then they should run `bundle exec rake clobber generate:app generate:stuff`.  That solves the upgrade problem entirely.

Reply:
https://github.com/rspec/rspec-rails/commit/c68b5680de1e6828caa152f1485e889c09e5cf4b#diff-94efcbcbb995cd8f60c1956dbb889ec2af5999f3aef85e686278dd96b9a3be16R21

A competitive commit coincidentally implementing the approach I suggest.

> Why does Rails 6 have a preferential position over Rails 7? What happens if I generate the app on Rails 7 and move to Rails 6.

There is no preferential approach. If you were on Rails 7, generated those specs with `index` with `<p><strong>`. 
If you move to Rails 6 after, `index_spec` that will now start asserting that `index` should have `tr>td`. But it doesn't, it still has `<p><strong>`.

There seem to be a massive misunderstanding on this.

Reply:
There is no misunderstanding.  You've implemented a version of the alternative that I described [here](https://github.com/rspec/rspec-rails/pull/2552#discussion_r781685831).  Specifically:

> So either we duplicate the logic in the spec (which seems kind of pointless) or we put it in a single variable here.

You've duplicated the logic from the index generator in the previously static scaffold spec.  Note there is no mention of such a change in [this comment](https://github.com/rspec/rspec-rails/pull/2552#discussion_r781787417).  As I noted, the issue is we have dynamically generated things (index specs) and one statically generated thing (scaffold spec).  Without the modification of the scaffold spec your specs would fail.

To me it seems obvious that duplicating the logic in two places (as you do in the referenced commit) is less ideal, but it's not a hill I want to die on.  If there's a consensus that this approach is "better" I'm fine with that.  As noted elsewhere, my goal is to get things working.

Keep in mind that regardless of this spec's behavior, switching Rails versions without regenerating the smoke app would count as a bad idea.  There's no guarantee that the generated app would i) run at all or ii) run consistently because there may be Rails version dependent changes.  My earlier comment about blowing away and regenerating the Rails app when switching Rails versions still holds.